rust/sealevel/programs/build-programs.sh

```diff
@@ -0,0 +1,93 @@
+#!/bin/bash
+
+# The script:
+# - uses the desired version of the solana cli for building programs
+# - resets the solana cli to the desired version after building programs
+# - builds the programs, accepting an arg for what types of programs to build
+# - if no arg is given, builds all programs
+
+# The first argument is the type of program to build
+PROGRAM_TYPE="${1:-all}"
+
+SOLANA_CLI_VERSION_FOR_BUILDING_PROGRAMS="1.14.20"
+
+# The paths to the programs
+CORE_PROGRAM_PATHS=("mailbox" "ism/multisig-ism-message-id" "validator-announce" "hyperlane-sealevel-igp")
+TOKEN_PROGRAM_PATHS=("hyperlane-sealevel-token" "hyperlane-sealevel-token-collateral" "hyperlane-sealevel-token-native")
+
+build_program () {
+    PROGRAM_PATH=$1
+    PREV_DIR=$(pwd)
+    log "Building $PROGRAM_PATH"
+    cd $PROGRAM_PATH
+    cargo build-sbf
+    cd $PREV_DIR
+}
+
+build_programs () {
+    # The list of programs to build
+    PROGRAM_PATH_LIST=()
+
+    BUILD_ALL="false"
+
+    if [ $PROGRAM_TYPE == "all" ]; then
+        BUILD_ALL="true"
+    fi
+
+    log "Building programs of type: $PROGRAM_TYPE"
+
+    if [ $PROGRAM_TYPE == "token" ] || [ $BUILD_ALL == "true" ] ; then
+        PROGRAM_PATH_LIST+=("${TOKEN_PROGRAM_PATHS[@]}")
+    fi
+
+    if [ $PROGRAM_TYPE == "core" ] || [ $BUILD_ALL == "true" ] ; then
+        PROGRAM_PATH_LIST+=("${CORE_PROGRAM_PATHS[@]}")
+    fi
+
+    log "Building programs: ${PROGRAM_PATH_LIST[@]}"
+
+    # Build the programs
+    for PROGRAM_PATH in "${PROGRAM_PATH_LIST[@]}"
+    do
+        build_program $PROGRAM_PATH
+    done
+}
+
+get_current_solana_cli_version () {
+    # `solana --version` expected output is:
+    #    solana-cli 1.18.18 (src:83047136; feat:4215500110, client:SolanaLabs)
+    # So we can use awk to get the second field
+    echo $(solana --version | awk '{print $2}')
+}
+
+set_solana_cli_version () {
+    NEW_VERSION=$1
+
+    sh -c "$(curl -sSfL https://release.solana.com/v$NEW_VERSION/install)"
+}
+
+log () {
+    echo "#### $@"
+}
+
+main () {
```

Comment:
maybe worth adding a cleanup helper and set a trap to ensure the cli version gets sack back to the original value at the end

```bash
cleanup () {
    # Only reset if we changed the version in the first place
    if [ "$SOLANA_CLI_VERSION_AT_START" != "$SOLANA_CLI_VERSION_FOR_BUILDING_PROGRAMS" ] && \
       [ ! -z "$SOLANA_CLI_VERSION_AT_START" ]; then
        log "Resetting Solana CLI version back to $SOLANA_CLI_VERSION_AT_START..."
        set_solana_cli_version $SOLANA_CLI_VERSION_AT_START
    fi
}
```

and before line 78 we have
```bash
trap cleanup EXIT
```

Reply:
good call, done